File: concrete/src/main/java/pcb/uwu/amount/quantity/Seconds.java
package pcb.uwu.amount.quantity;

import pcb.uwu.amount.composite.electromagnetism.Coulombs;
import pcb.uwu.amount.composite.termodynamics.Joules;
import pcb.uwu.amount.composite.termodynamics.Power;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.quantity.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.composite.termodynamics.WattUnit.WATT;
import static pcb.uwu.unit.quantity.AmpereUnit.AMPERE;
import static pcb.uwu.unit.quantity.SecondUnit.SECOND;

public class Seconds extends Time {

	// region constructors

	public Seconds(Number value) {
		super(value, SECOND);
	}

	public Seconds(Number value, Magnitude magnitude) {
		super(value, magnitude, SECOND);
	}

	public Seconds(String value) {
		super(value, SECOND);
	}

	public Seconds(String value, Magnitude magnitude) {
		super(value, magnitude, SECOND);
	}

	public Seconds(BigDecimal value) {
		super(value, SECOND);
	}

	public Seconds(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, SECOND);
	}

	public Seconds(BigDecimalAmount amount) {
		super(amount, SECOND);
	}

	public Seconds(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, SECOND);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Seconds plus(UnitAmount<TimeUnit> other, MathContext mathContext) {
		return new Seconds(plusAmount(other, mathContext));
	}

	@Override
	public Seconds minus(UnitAmount<TimeUnit> other, MathContext mathContext) {
		return new Seconds(minusAmount(other, mathContext));
	}

	@Override
	public Seconds multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Seconds(multipliedByScalar(other, mathContext));
	}

	@Override
	public Seconds dividedBy(BigDecimal other, MathContext mathContext) {
		return new Seconds(dividedByScalar(other, mathContext));
	}

	// endregion

	// region composite

	public Coulombs multipliedBy(ElectricCurrent electricCurrent, MathContext mathContext) {
		return new Coulombs(getAmount().multipliedBy(electricCurrent.getAmountIn(AMPERE), mathContext));
	}

	public Joules multipliedBy(Power power, MathContext mathContext) {
		return new Joules(getAmount().multipliedBy(power.getAmountIn(WATT), mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Seconds.kt
package pcb.uwu.amount.base

import pcb.uwu.amount.derived.electromagnetism.Coulombs
import pcb.uwu.amount.derived.optics.LuminousFlux
import pcb.uwu.amount.derived.optics.Talbots
import pcb.uwu.amount.derived.thermodynamics.Joules
import pcb.uwu.amount.derived.thermodynamics.Power
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.AmpereUnit
import pcb.uwu.unit.base.SecondUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.optics.LumenUnit
import pcb.uwu.unit.derived.termodynamics.WattUnit

class Seconds : Time
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, SecondUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, SecondUnit)

    // region UnitAmount

    override fun plus(time: UnitAmount<TimeUnit>) =
        Seconds(this.amount + (time to SecondUnit).amount)

    override fun minus(time: UnitAmount<TimeUnit>) =
        Seconds(this.amount - (time to SecondUnit).amount)

    override fun times(scalar: Number) =
        Seconds(this.amount * scalar)

    override fun div(scalar: Number) =
        Seconds(this.amount / scalar)

    // endregion

    // region composition

    override fun times(electricCurrent: ElectricCurrent) =
        Coulombs(this.amount * (electricCurrent to AmpereUnit).amount)

    override fun times(power: Power) =
        Joules(this.amount * (power to WattUnit).amount)

    override fun times(luminousFlux: LuminousFlux) =
        Talbots(this.amount * (luminousFlux to LumenUnit).amount)

    // endregion
}